tasks/actorstate/datacap/balance.go

```diff
@@ -0,0 +1,137 @@
+package datacap
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+	logging "github.com/ipfs/go-log/v2"
+	"go.opentelemetry.io/otel"
+	"go.uber.org/zap"
+
+	"github.com/filecoin-project/lily/chain/actors/builtin/datacap"
+	"github.com/filecoin-project/lily/model"
+	datacapmodel "github.com/filecoin-project/lily/model/actors/datacap"
+	"github.com/filecoin-project/lily/tasks/actorstate"
+)
+
+var log = logging.Logger("lily/tasks/datacap")
+
+type BalanceExtractor struct{}
+
+func (BalanceExtractor) Extract(ctx context.Context, a actorstate.ActorInfo, node actorstate.ActorStateAPI) (model.Persistable, error) {
+	log.Debugw("extract", zap.String("extractor", "BalanceExtractor"), zap.Inline(a))
+	ctx, span := otel.Tracer("").Start(ctx, "BalancesExtractor.Extract")
+	defer span.End()
+	if span.IsRecording() {
+		span.SetAttributes(a.Attributes()...)
+	}
+
+	ec, err := NewBalanceExtractionContext(ctx, a, node)
+	if err != nil {
+		return nil, err
+	}
+
+	var balances datacapmodel.DataCapBalanceList
+
+	// should only be true on the upgrade boundary
```

Comment:
Maybe add more context that it's the nv16 to nv17 upgrade? (Actually not sure myself if that's what you mean here).

Also couldn't there be a case where a datacap actor is only created for e.g. 20000 epochs after a network upgrade?

Reply:
Datacap actor is a singleton. It's only true on the upgrade boundary because the actor was introduced in nv17 and didn't exist in nv16.